magenta-lib/src/main/scala/magenta/deployment_type/BucketParameters.scala

```diff
@@ -28,28 +28,59 @@ trait BucketParameters {
     """The SSM key used to lookup the bucket name for uploading distribution artifacts.""".stripMargin
   ).default(defaultSsmKeyParamDefault)
 
+  val bucketSsmKeyStageParam = Param[Map[String, String]](
+    "bucketSsmKeyStageParam",
+    """
+      |Like bucketSsmKeyParam, but with the ability to configure by stage:
+      |
+      |```yaml
+      |  CODE: some-ssm-path-for-code
+      |  PROD: some-ssm-path-for-prod
+      |```
+      |""".stripMargin
+  ).default(Map.empty)
+
   def getTargetBucketFromConfig(
       pkg: DeploymentPackage,
       target: DeployTarget,
       reporter: DeployReporter
   ): Bucket = {
+
+    def bySsm(): Bucket = {
+      val stage = target.parameters.stage.name
+      val ssmKeyByStage = bucketSsmKeyStageParam(pkg, target, reporter)
+      val stageKey = ssmKeyByStage.get(stage)
+
+      if (ssmKeyByStage.nonEmpty && stageKey.isEmpty) {
+        reporter.fail(
+          s"Unable to determine bucket to deploy to: bucketSsmKeyStageParam is set but no mapping was found for stage '$stage'."
+        )
+      }
+
+      BucketBySsmKey(
+        stageKey.getOrElse(bucketSsmKeyParam(pkg, target, reporter))
+      )
+    }
+
     val bucketSsmLookup = bucketSsmLookupParam(pkg, target, reporter)
-    val maybeExplicitBucket = bucketParam.get(pkg)
+    val explicitBucket = bucketParam.get(pkg)
 
-    val bucket = (bucketSsmLookup, maybeExplicitBucket) match {
-      // Default to looking up target bucket from SSM
-      case (_, None) =>
-        BucketBySsmKey(bucketSsmKeyParam(pkg, target, reporter))
-      case (false, Some(explicitBucket)) =>
+    // The behaviour here is *very* counter-intuitive; even if bucketSsmLookup=false
```

Comment:
I think what's missing here is the default value of `bucketSsmLookup` is `false` and the default behaviour we want is to use `bucketSsmLookup` - so we are confusingly using the signal that `bucketSsmLookup` hasn't been changed to perform that default behaviour!

So should the default value of `bucketSsmLookup` be `true` and then we can remove some of the confusing logic here.

Reply:
Or after talking with @nicl ... maybe we can just remove the param?

Reply:
👍 will do this in a follow-up once I've confirmed it is okay. But I think we should be safe to remove the `bucketSsmLookup` parameter completely.

Note, 'remove' here means ignoring it and reporting a warning whenever it is set. We can't remove it unfortunately as RR errors on unrecognised fields in a riff-raff.yaml file.